spec/models/split_spec.rb

```diff
@@ -208,4 +208,162 @@
       expect(described_class.active(as_of: Time.zone.now)).not_to include(split)
     end
   end
+
+  describe ".with_feature_incomplete_knockouts_for" do
+    let(:app) { FactoryBot.create(:app) }
+    let(:app_build) { app.define_build(built_at: Time.zone.now, version: "1.0") }
+
+    it "respects existing selects" do
+      split = FactoryBot.create(:split, feature_gate: false)
+
+      result = Split.select(:name).with_feature_incomplete_knockouts_for(app_build).find(split.id)
+
+      expect(result).not_to be_feature_incomplete
+      expect(result).to respond_to(:name)
+      expect(result).not_to respond_to(:finished_at)
+    end
+
+    it "isn't feature_incomplete for non-feature gates" do
+      split = FactoryBot.create(:split, feature_gate: false)
+      expect(Split.with_feature_incomplete_knockouts_for(app_build).find(split.id)).not_to be_feature_incomplete
+    end
+
+    it "returns readonly records" do
+      split = FactoryBot.create(:split, feature_gate: false)
+      expect(Split.with_feature_incomplete_knockouts_for(app_build).find(split.id)).to be_readonly
+    end
+
+    it "isn't feature_incomplete for feature gates with a feature completion" do
+      split = FactoryBot.create(:split, feature_gate: true)
+      FactoryBot.create(:app_feature_completion, app: app, version: "1.0", split: split)
+      expect(Split.with_feature_incomplete_knockouts_for(app_build).find(split.id)).not_to be_feature_incomplete
+    end
+
+    it "is feature_incomplete for feature gates with no feature completion" do
+      split = FactoryBot.create(:split, feature_gate: true)
+      expect(Split.with_feature_incomplete_knockouts_for(app_build).find(split.id)).to be_feature_incomplete
+    end
+
+    it "is backed by AppFeatureCompletion.satisfied_by" do
+      allow(AppFeatureCompletion).to receive(:satisfied_by).and_call_original
+
+      Split.with_feature_incomplete_knockouts_for(app_build)
+
+      expect(AppFeatureCompletion).to have_received(:satisfied_by).with(app_build)
```

Comment:
This doesn't _quite_ prove that it was effectively merged into the query, just that the code was executed. But aside from fully testing the integration the only other thing here would be to inspect the SQL or something.

Reply:
(similar arguments would apply elsewhere too, not sure how far we would want to take that.)

Reply:
Yeah, for sure not proof. The testing strategy generally for composed query logic is attempting to satisfy occam's razor that "if it works like this and calls that thing that is purpose built to create that behavior, it probably works like this because it is using that thing" - if you don't feel that we're comfortably hitting that balance, I think that'd be where the heat would be.

Reply:
There may be a way that I just haven't discovered of cracking open an AR relation and comparing its innards with something else, which would enable something like this:

```ruby
query = Split.with_feature_incomplete_knockouts_for(app_build)
query_part = AppFeatureCompletion.satisifed_by(app_build)
expect(query).to have_incorporated(query_part)
```

I've also controversially done something like this in the past:

```ruby
expect(query.to_sql).to include(<actual subquery SQL>)
```

But I'm not proposing that we do that here. I'd say for now I'm comfortable with the balance we're hitting, even if it's not an ideal state.

Reply:
Yeah, I think that's a good strategy too, and IIRC I've done it before also. It'd probably be a bit more brittle (irrelevant if this is low-rate-of-change), but stronger confidence for sure. I'm happy with the confidence level as-is though, so yeah, let's rock it. Thanks for the deep dive.